### ISSUE TITLE
Authentication tokens should expire after a given amount of time

### ISSUE THREAD

**Issue by didair** [spaceuptech/space-cloud#204]
### The problem faced currently?
Right now the token sent with the authentication payload doesn't have a parameter for when it expires.

### How can we solve it?
Implement a DateTime property in the JWT token and make sure that the authentication also gets invalid when it expires on the back end.

>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

**Comment by YourTechBud** [comment]
Makes sense. We can do this. However it shouldn't be the default behaviour. Maybe adding some configuration can activate the lease period for the token.

@didair i have a question here. Should the token automatically be refreshed by the client? Or is the user responsible to do that? Automatic refresh won't work for scenario using 3rd party sign on methods (google, fb, ldap, etc.). Any suggestions?

**Comment by didair** [comment]
@YourTechBud no the client shouldnt refresh the token automatically but there should definitely be a built in function in the client to do that. I think the behaviour should be decided by the app and not Space Cloud.

I also like the idea that this feature should be a configuration.

**Comment by YourTechBud** [comment]
That makes sense. Also, the token cannot be refreshed after it has expired right? The user will have to manually sign in again?

**Comment by didair** [comment]
well.. No. That's a hard question but in this case, no. If the token expires you need to make a manual sign in. I think making it work as close as possible to the current functionality is good.

If you look at how ADAL works you have first an idToken that represent the current authentication session which normally has a longer expiry time. With the id token you can decide if the user is signed in or not and if it's expired but you have it in memory you can renew it. You then have another token for making API calls that handles API scopes etc. I don't think having two separate tokens in this case is applicable for Space Cloud but ADAL might be a good place to look at for inspiration.

**Comment by YourTechBud** [comment]
@didair So if I understand this correctly the token renewal process can only be initiated by the client (obviously). The token will be renewed when one of the following conditions holds true.
- the token has not yet expired. This will let the client extend the expiry of the token.
- the token has expired but is still held by the client. (Not sure about this one)

Am I missing something here?

**Comment by thienpow** [comment]
just a suggestion, maybe can use Redis to store the jwt token temporary, this would allow a Redis look for expired jwt token to be flushed. and it will speed up client to retrieve existing jwt token if it's not expired. also might as well no need pass the token into a thousands functions in the code... i was drilling into the codes, seeing the token keep being passed into other functions, but most of them not being used... and it's used by the very last function in the tree... if redis is being used, maybe the code can be cleaner.

another reason to use redis, is bcos the token should be identical to each browser or app... currently the token is able to be consumed everywhere, for existing logic, example if i login from chrome, i can steal the token and use it in a custom app or use it at another browser without login again.  with redis, we could have recorded the token like "uid", "sourceApp", "sourceAddress", "token", "expiryDT"... with these, we could make token hijack being harder to hack.

**Comment by YourTechBud** [comment]
@thienpow those are some great points.

```
 i was drilling into the codes, seeing the token keep being passed into other functions, but most of them not being used... and it's used by the very last function in the tree
```
Yeah, we need to get that fix.

```
with redis, we could have recorded the token like "uid", "sourceApp", "sourceAddress", "token", "expiryDT"... with these, we could make token hijack being harder to hack.
```
Can't we just store these in the token itself?

Also, regarding storing sessional state in redis, thats a great idea. However, I'm a bit hesitate to add a dependency on redis. Maybe we can solve this without having to store stuff in Redis?

Regarding expiry and renewal of tokens, user management module of space-cloud isn't intended to be used in production. The user is expected to generate and renew tokens using `remote services`.

I'm letting this issue be since its such a common task that we are considering making an official `remote service ` for IAM.

**Comment by YourTechBud** [comment]
Fixed in `v0.19.0`
